controller/src/util/instance_action.rs

```diff
@@ -472,17 +474,41 @@ pub async fn handle_instance_change(
         .await?
     }
 
-    let nodes_to_add = nodes_to_act_on
+    // Check how many nodes already have pods
+    let num_nodes_already_with_pods: i32 = nodes_to_act_on
         .iter()
-        .filter_map(|(node, context)| {
-            if ((context.action) == PodAction::Add) | ((context.action) == PodAction::RemoveAndAdd)
-            {
-                Some(node.to_string())
+        .filter_map(|(_, context)| {
+            if (context.action) == PodAction::NoAction {
```

Comment:
i don't remember ... does NoAction equate to a pod is there?

Reply:
are we trying to get the used capacity here?  or are we trying to get the number of brokers that have been deployed?

Reply:
Instance.deviceUsage will include broker pods AND pods that people have tried to schedule.

If we query pods for the labels our controller adds, we will only get brokers.

Reply:
Hmm thats an interesting point. So you're saying that this scenario could happen
1. Operator creates a Configuration with capacity of 5 in a two worker node cluster
2. Controller deploys 2 brokers -> Agents reserve 2 of the 5 `deviceUsage` slots
3. Operator deploys 2 pods requesting that Instance as a resource limit
4. Agent reserves 2 slots for those pods
Now 4 of 5 slots have been taken but the Akri Controller only deployed 2 brokers. Is this the behavior we want? It seems that the controller should check how many pods are using the instance, whether Akri deployed broker or not. 

But to answer your question: I was using the number of nodes with `PodAction:NoAction` as the number of brokers already deployed because [when determining an action for a pod](https://github.com/deislabs/akri/blob/c80e73d049fde9b57c8aa8308a248e7afe7002c2/controller/src/util/instance_action.rs#L455), running ones keep this status. Pods that should be removed also get the [NoAction](https://github.com/deislabs/akri/blob/c80e73d049fde9b57c8aa8308a248e7afe7002c2/controller/src/util/instance_action.rs#L412) label which may not be modified if there is an early return from [`determine_action_for_pod`](https://github.com/deislabs/akri/blob/c80e73d049fde9b57c8aa8308a248e7afe7002c2/controller/src/util/instance_action.rs#L163). 

TLDR: this need to change and like you said doesnt take into account the other pods that have been scheduled to a resource that arent brokers. So should we use the number of taken `deviceUsage` slots as the capacity?

Reply:
That is the scenario I was thinking about.  I do think that Instance.deciceUsage should be the source of truth.

Part of me thinks that the controller shouldn't worry about capacity...just create a pod for each node in Instance.nodes and let the scheduler deal with the rest.   As a result, there will be some Pending pods, but that just ensures that we continually have HA.

But I'm not sure I have thought that completely through.

Reply:
Alternatively, we could remove the node affinity and just create Instance.caoacity pods....what is scheduled and what is pending will depend on allocate (which will control operator deployed pods as well)

Reply:
The second s probably more efficient if there are lots and lots of nodes.

And probably better for us to get out of the node choosing game.

Reply:
But is a bigger change in behavior.

Reply:
In this case, would the controller even need to monitor instance changes? Would the scheduler also be in charge of bringing down pods that are `Pending`. I agree that it makes sense to move away from the controller choosing the node.

Reply:
I wonder if we should have controller check capacity (based on deviceUsage used slots) for now, especially if we want to set default capacity to 1, since there is a higher chance of the pending pods case. Or we can make the controller better at bringing down pending pods.

Reply:
i think we should stop worrying about Pending.

i like this for strategies:

* **CapacityDaemonSet** - this is what we have now.  if capacity=5 and there are 10 nodes that can see an instance, we create 10 pods with node-affinity, 5 end up Pending.
* **CapacityDeployment** - this would be new (and should be the eventual default) ... if capacity=5 and there are 10 nodes that can see an instance, a deployment with 5 replicas is created using resources.limit and PLACEHOLDER for targeting (no node-affinity, k8s scheduler owns scheduling)